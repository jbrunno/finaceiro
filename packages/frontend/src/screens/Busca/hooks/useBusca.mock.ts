import { faker } from '@faker-js/faker/locale/pt_BR';

export const clientes = {
  items: [
    {
      node: {
        __typename: 'Cliente',
        cpfCnpj: faker.random.numeric(11),
        falecido: faker.datatype.boolean(),
        id: faker.datatype.uuid(),
        nome: faker.name.fullName(),
        negativado: faker.datatype.boolean(),
        nomeSocial: faker.name.firstName(),
      },
    },
  ],
  pageInfo: {
    hasNextPage: faker.datatype.boolean(),
    totalItems: faker.datatype.number(10),
  },
};
