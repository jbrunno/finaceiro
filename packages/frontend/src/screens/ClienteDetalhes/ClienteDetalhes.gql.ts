import { ClienteContatosEmailTable } from '@/shared/components/ClienteContato/components/ClienteContatosEmailTable/ClienteContatosEmailTable';
import { ClienteContatosEmail } from '@/shared/components/ClienteContato/components/ClienteContatosEmailTable/ClienteContatosEmailTable.gql';
import { ClienteContatosEnderecoTable } from '@/shared/components/ClienteContato/components/ClienteContatosEnderecoTable/ClienteContatosEnderecoTable';
import { ClienteContatosEndereco } from '@/shared/components/ClienteContato/components/ClienteContatosEnderecoTable/ClienteContatosEnderecoTable.gql';
import { ClienteContatosTelefoneTable } from '@/shared/components/ClienteContato/components/ClienteContatosTelefoneTable/ClienteContatosTelefoneTable';
import { ClienteContatosTelefone } from '@/shared/components/ClienteContato/components/ClienteContatosTelefoneTable/ClienteContatosTelefoneTable.gql';
import { ClienteInfoCard } from '@/shared/components/ClienteInfoCard/ClienteInfoCard';
import { ClienteInfoData } from '@/shared/components/ClienteInfoCard/ClienteInfoCard.gql';
import { FollowUpCard } from '@/shared/components/FollowUpCard/FollowUpCard';
import {
  ClienteFollowUps,
  SituacoesFollowUp,
} from '@/shared/components/FollowUpCard/FollowUpCard.gql';
import { TituloCard } from '@/shared/components/TituloCard/TituloCard';
import { TituloQuestionario } from '@/shared/components/TituloCard/TituloCard.gql';
import { gql } from '@apollo/client';

export const CLIENTE_DETALHES_QUERY = gql`
  query Cliente($clienteId: String!) {
    cliente(id: $clienteId) {
      ...CLIENTE_INFO_FRAGMENT
    }
  }

  ${ClienteInfoCard.fragments.cliente}
`;

export const CLIENTE_CONTATO_EMAILS_QUERY = gql`
  query Emails($id: String!, $pagination: PaginationInput!) {
    emails(id: $id, pagination: $pagination) {
      items {
        node {
          ...CLIENTE_CONTATO_EMAIL_FRAGMENT
        }
      }
      pageInfo {
        totalItems
        hasNextPage
      }
    }
  }
  ${ClienteContatosEmailTable.fragments.emails}
`;

export const CLIENTE_CONTATO_TELEFONES_QUERY = gql`
  query Telefones($id: String!, $pagination: PaginationInput!) {
    telefones(id: $id, pagination: $pagination) {
      items {
        node {
          ...CLIENTE_CONTATO_TELEFONE_FRAGMENT
        }
      }
      pageInfo {
        totalItems
        hasNextPage
      }
    }
  }
  ${ClienteContatosTelefoneTable.fragments.telefones}
`;

export const CLIENTE_CONTATO_ENDERECOS_QUERY = gql`
  query Enderecos($id: String!, $pagination: PaginationInput!) {
    enderecos(id: $id, pagination: $pagination) {
      items {
        node {
          ...CLIENTE_CONTATO_ENDERECO_FRAGMENT
        }
      }
      pageInfo {
        totalItems
        hasNextPage
      }
    }
  }
  ${ClienteContatosEnderecoTable.fragments.enderecos}
`;

export const CLIENTE_TITULOS_QUERY = gql`
  query Titulos($clienteId: String!) {
    titulos(id: $clienteId) {
      items {
        node {
          id
          assinatura
          tituloQuestionario {
            ...CLIENTE_TITULO_FRAGMENT
          }
        }
      }
    }
  }

  ${TituloCard.fragments.tituloQuestionario}
`;

export const CLIENTE_FOLLOWUPS_QUERY = gql`
  query FollowUps($clienteId: String!, $pagination: PaginationInput!) {
    followUps(id: $clienteId, pagination: $pagination) {
      items {
        node {
          ...CLIENTE_FOLLOWUPS_FRAGMENT
        }
      }
      pageInfo {
        totalItems
        hasNextPage
      }
    }
  }

  ${FollowUpCard.fragments.followps}
`;

export const SITUACOES_FOLLOWUP_QUERY = gql`
  query SituacoesFollowUp($keyword: String, $pagination: PaginationInput!) {
    situacoesFollowUp(keyword: $keyword, pagination: $pagination) {
      items {
        node {
          ...SITUACOES_FOLLOWUPS_FRAGMENT
        }
      }
      pageInfo {
        totalItems
        hasNextPage
      }
    }
  }

  ${FollowUpCard.fragments.situacoes}
`;

export type PaginationInput = {
  pageNumber: number;
  pageSize: number;
};

type PaginationData<T> = {
  items: Array<{
    node: T;
  }>;
  pageInfo: {
    totalItems: number;
    hasNextPage: boolean;
  };
};

export type ClienteDetalhesQueryVariables = {
  clienteId: string;
};

export type ClienteDetalhesData = {
  cliente: ClienteInfoData;
};

export type ClienteContatoQueryVariables = {
  id: string;
  pagination: PaginationInput;
};

export type ClienteContatoTelefoneData = {
  telefones: PaginationData<ClienteContatosTelefone>;
};

export type ClienteContatoEnderecoData = {
  enderecos: PaginationData<ClienteContatosEndereco>;
};

export type ClienteContatoEmailsData = {
  emails: PaginationData<ClienteContatosEmail>;
};

export type TitulosQueryVariables = {
  clienteId: string;
};

export type TituloData = {
  titulos: {
    items: Array<{
      node: {
        id: string;
        assinatura: boolean;
        tituloQuestionario: TituloQuestionario;
      };
    }>;
  };
};

export type ClienteFollowUpsQueryVariables = {
  clienteId: string;
  pagination: PaginationInput;
};

export type SituacoesFollowUpsQueryVariables = {
  keyword?: string;
  pagination: PaginationInput;
};

export type ClienteFollowUpsData = {
  followUps: PaginationData<ClienteFollowUps>;
};

export type SituacoesFollowUpData = {
  situacoesFollowUp: PaginationData<SituacoesFollowUp>;
};
