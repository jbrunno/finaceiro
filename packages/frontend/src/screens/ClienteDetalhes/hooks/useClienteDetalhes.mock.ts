import {
  ClienteSexoEnum,
  ClienteTipoEnum,
  ContatoClassificacaoEnum,
  ContatoMarcadorEnum,
} from '@/shared/constants';
import { faker } from '@faker-js/faker/locale/pt_BR';

const cbo = {
  id: faker.datatype.uuid(),
  codigo: faker.random.numeric(3),
  ocupacao: faker.name.jobTitle(),
};

export const cliente = {
  __typename: 'Cliente',
  id: faker.datatype.uuid(),
  nome: faker.name.fullName(),
  nomeSocial: faker.name.firstName(),
  documento: faker.random.numeric(11),
  sexo: ClienteSexoEnum.MASCULINO,
  dataNascimento: faker.date.birthdate().toISOString(),
  cbo,
  falecido: faker.datatype.boolean(),
  negativado: faker.datatype.boolean(),
  tipo: ClienteTipoEnum.ESTRANGEIRA,
};

const tituloQuestionario = {
  __typename: 'TituloQuestionario',
  siteRegistro: faker.internet.url(),
  tipoConta: faker.finance.accountName(),
  contaAnteriorFalencia: faker.datatype.boolean(),
  tipoAtivoInvestido: faker.commerce.product(),
  valorDepositos: faker.finance.amount(0, 10000),
  observacao: faker.lorem.text(),
};

export const titulos = {
  items: [
    {
      node: {
        __typename: 'Titulos',
        id: faker.datatype.uuid(),
        assinatura: faker.datatype.boolean(),
        tituloQuestionario,
      },
    },
  ],
};

const pageInfo = {
  hasNextPage: faker.datatype.boolean(),
  totalItems: faker.datatype.number(10),
};

export const emails = {
  items: [
    {
      node: {
        __typename: 'ContatoEmail',
        id: faker.datatype.uuid(),
        nome: faker.name.fullName(),
        email: faker.internet.email(),
        marcador: ContatoMarcadorEnum.PAI,
        classificacao: ContatoClassificacaoEnum.BOM,
      },
    },
  ],
  pageInfo,
};

export const enderecos = {
  items: [
    {
      node: {
        __typename: 'ContatoEndereco',
        id: faker.datatype.uuid(),
        nome: faker.name.fullName(),
        logradouro: faker.address.streetAddress(),
        numero: faker.address.buildingNumber(),
        bairro: faker.address.street(),
        cidade: faker.address.cityName(),
        uf: faker.address.stateAbbr(),
        cep: faker.address.zipCode(),
        complemento: faker.address.secondaryAddress(),
        marcador: ContatoMarcadorEnum.PAI,
        classificacao: ContatoClassificacaoEnum.BOM,
        pais: {
          id: faker.datatype.uuid(),
          descricao: faker.address.country(),
        },
      },
    },
  ],
  pageInfo,
};

export const telefones = {
  items: [
    {
      node: {
        __typename: 'ContatoTelefone',
        id: faker.datatype.uuid(),
        nome: faker.name.fullName(),
        numero: faker.phone.number(),
        whatsapp: faker.datatype.boolean(),
        marcador: ContatoMarcadorEnum.PAI,
        classificacao: ContatoClassificacaoEnum.BOM,
      },
    },
  ],
  pageInfo,
};

export const clienteFollowUps = {
  items: [
    {
      node: {
        __typename: 'FollowUp',
        id: faker.datatype.uuid(),
        dataRegistro: faker.date.past().toISOString(),
        usuario: {
          id: faker.datatype.uuid(),
          nome: faker.name.firstName(),
          username: faker.internet.userName(),
          turno: faker.datatype.number({ min: 4, max: 8 }),
          cargaHoraria: faker.random.numeric(4),
        },
        situacao: {
          id: faker.datatype.uuid(),
          descricaoObrigatoria: faker.datatype.boolean(),
          codigo: faker.datatype.number({ min: 4, max: 8 }),
          nome: faker.name.firstName(),
          tempoTrava: 0,
          textoPadrao: faker.lorem.sentence(),
        },
        descricao: faker.lorem.sentence(),
      },
    },
  ],
  pageInfo,
};

export const situacoesFollowUp = {
  items: [
    {
      node: {
        __typename: 'Situacao',
        id: faker.datatype.uuid(),
        descricaoObrigatoria: faker.datatype.boolean(),
        codigo: faker.datatype.number({ min: 4, max: 8 }),
        nome: faker.name.firstName(),
        tempoTrava: 0,
        textoPadrao: faker.lorem.sentence(),
      },
    },
  ],
  pageInfo,
};
