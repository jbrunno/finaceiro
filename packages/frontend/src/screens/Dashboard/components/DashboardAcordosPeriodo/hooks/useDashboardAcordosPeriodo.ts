import { useState, ChangeEvent } from 'react';
import { crmPeriods } from '@frontend/utils';
import { PeriodsEnum } from '@frontend/types';
import { ProducaoAcordosSintetizadoData } from '../DashboardAcordosPeriodo.gql';

const periods = [
  {
    label: 'Hoje',
    value: crmPeriods(PeriodsEnum.TODAY).toISOString(),
  },
  {
    label: 'Dia anterior',
    value: crmPeriods(PeriodsEnum.YESTERDAY).toISOString(),
  },
  {
    label: 'Últimos 7 dias',
    value: crmPeriods(PeriodsEnum.LAST7DAYS).toISOString(),
  },
  {
    label: 'Últimos 15 dias',
    value: crmPeriods(PeriodsEnum.LAST15DAYS).toISOString(),
  },
  {
    label: 'Mês atual (até hoje)',
    value: crmPeriods(PeriodsEnum.THISMONTH).toISOString(),
  },
  {
    label: 'Mês anterior',
    value: crmPeriods(PeriodsEnum.LASTMONTH).toISOString(),
  },
  {
    label: 'Últimos 6 meses',
    value: crmPeriods(PeriodsEnum.LAST6MONTHS).toISOString(),
  },
];

export type UseDashboardAcordosPeriodoProps = {
  producaoAcordosSintetizado?: ProducaoAcordosSintetizadoData;
  onChangePeriod: (period: string) => void;
};

export const useDashboardAcordosPeriodo = ({
  producaoAcordosSintetizado,
  onChangePeriod,
}: UseDashboardAcordosPeriodoProps) => {
  const acordosInfoProgress = [
    {
      title: 'Acordos pagos',
      qty: producaoAcordosSintetizado?.quantidadeAcordosPagos || 0,
    },
    {
      title: 'Acordos quebrados',
      qty: producaoAcordosSintetizado?.quantidadeAcordosQuebrados || 0,
    },
    {
      title: 'Acordos com parcelas a vencer',
      qty: producaoAcordosSintetizado?.quantidadeParcelasAVencer || 0,
    },
  ];

  const acordosInfoProgressTotal = acordosInfoProgress.reduce(
    (acc, curr) => acc + curr.qty,
    0,
  );

  const [period, setPeriod] = useState<string>(periods[4].value);

  const handleChange = (event: ChangeEvent<HTMLInputElement>) => {
    setPeriod(event.target.value);
    onChangePeriod(event.target.value);
  };

  return {
    handleChange,
    period,
    periods,
    acordosInfoProgressTotal,
    acordosInfoProgress,
  };
};
