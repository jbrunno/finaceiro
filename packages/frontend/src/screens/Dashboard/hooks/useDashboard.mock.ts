import { faker } from '@faker-js/faker';
import { ProducaoAcordosSintetizadoData } from '../components/DashboardAcordosPeriodo/DashboardAcordosPeriodo.gql';
import { IcmData } from '../components/DashboardIcm/DashboardIcm.gql';

type WithTypename<T> = T & { __typename: string };

export const icm: WithTypename<IcmData> = {
  __typename: 'Icm',
  porcentagemTotal: faker.datatype.float({ max: 99 }),
  porcentagemEntrada: faker.datatype.float({ max: 99 }),
  valorEntrada: faker.datatype.number(1000),
  valorParcelasPagas: faker.datatype.number(12),
  porcentagemParcelasPagas: faker.datatype.float({ max: 99 }),
};

export const producaoAcordosSintetizado: WithTypename<ProducaoAcordosSintetizadoData> =
  {
    __typename: 'ProducaoAcordosSintetizado',
    acordosEPromessasDePagamento: faker.datatype.number(12),
    quantidadeAcordosPagos: faker.datatype.number(12),
    quantidadeAcordosQuebrados: faker.datatype.number(12),
    quantidadeParcelasAVencer: faker.datatype.number(12),
    valorTotalAcordosGerados: faker.datatype.number(3000),
    valorTotalAcordosPagos: faker.datatype.number(2000),
  };

export const producaoAcordosComparativo = {
  __typename: 'ProducaoAcordosComparativo',
  ...[
    {
      data: faker.datatype.datetime().toISOString(),
      quantidadeIndividual: faker.datatype.number(30),
      quantidadeMedia: faker.random.numeric(3),
    },
  ],
};
