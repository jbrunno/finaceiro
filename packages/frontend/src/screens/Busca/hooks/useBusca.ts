import { useLazyQuery } from '@frontend/hooks';
import { deepMerge } from '@frontend/utils';
import { useEffect, useState } from 'react';
import { useSearchParams } from 'react-router-dom';
import { BuscaData, BuscaQueryVariables, BUSCA_QUERY } from '../Busca.gql';

export const useBusca = () => {
  const [params] = useSearchParams();
  const searchParam = params.get('search');
  const [pageNumber, setPageNumber] = useState(1);
  const [data, setData] = useState<BuscaData | undefined>(undefined);
  const [getClientes, { data: clientesData, loading, fetchMore }] =
    useLazyQuery<BuscaData, BuscaQueryVariables>(BUSCA_QUERY);

  useEffect(() => {
    if (searchParam) {
      getClientes({
        keyword: searchParam || '',
        pagination: {
          pageNumber,
          pageSize: 10,
        },
      });
    }
  }, [searchParam]);

  useEffect(() => {
    if (clientesData && searchParam) {
      setData(clientesData);
    } else {
      setData(undefined);
    }
  }, [clientesData, searchParam]);

  const loadMore = () => {
    fetchMore({
      variables: {
        pagination: {
          pageNumber: pageNumber + 1,
          pageSize: 10,
        },
      },
      updateQuery: (prev, { fetchMoreResult }) => {
        if (!fetchMoreResult) return prev;

        return deepMerge(prev, fetchMoreResult);
      },
    });
    setPageNumber(pageNumber + 1);
  };

  return {
    data,
    loading,
    searchParam,
    loadMore,
  };
};
