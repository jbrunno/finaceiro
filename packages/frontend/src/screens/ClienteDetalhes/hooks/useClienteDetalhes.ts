import { DASHBOARD } from '@/router/routes';
import { useQuery } from '@apollo/client';
import { deepMerge } from '@frontend/utils';
import { useEffect, useState } from 'react';
import { useObjectState, useRoute } from '@frontend/hooks';
import { useParams } from 'react-router-dom';
import {
  ClienteContatoEmailsData,
  ClienteContatoEnderecoData,
  ClienteContatoQueryVariables,
  ClienteContatoTelefoneData,
  ClienteDetalhesData,
  ClienteDetalhesQueryVariables,
  CLIENTE_CONTATO_EMAILS_QUERY,
  CLIENTE_CONTATO_ENDERECOS_QUERY,
  CLIENTE_CONTATO_TELEFONES_QUERY,
  CLIENTE_DETALHES_QUERY,
  TitulosQueryVariables,
  CLIENTE_TITULOS_QUERY,
  TituloData,
  ClienteFollowUpsQueryVariables,
  ClienteFollowUpsData,
  SituacoesFollowUpData,
  SituacoesFollowUpsQueryVariables,
  SITUACOES_FOLLOWUP_QUERY,
  CLIENTE_FOLLOWUPS_QUERY,
} from '../ClienteDetalhes.gql';

const initialPageState = {
  emails: 1,
  telefones: 1,
  enderecos: 1,
};

type Contato = Partial<ClienteContatoEmailsData> &
  Partial<ClienteContatoTelefoneData> &
  Partial<ClienteContatoEnderecoData>;

export function useClienteDetalhes() {
  const { open: openDashboard, route } = useRoute(DASHBOARD);
  const [page, setPage] = useObjectState(initialPageState);
  const [contato, setContato] = useObjectState<Contato>({
    emails: undefined,
    telefones: undefined,
    enderecos: undefined,
  });
  const [situacoesKeyword, setSituacoesKeyword] = useState<
    string | undefined
  >();
  const { clienteId } = useParams();

  if (!clienteId) {
    openDashboard();
  }

  const { data, loading } = useQuery<
    ClienteDetalhesData,
    ClienteDetalhesQueryVariables
  >(CLIENTE_DETALHES_QUERY, {
    variables: {
      clienteId: clienteId || '',
    },
  });

  const emails = useQuery<
    ClienteContatoEmailsData,
    ClienteContatoQueryVariables
  >(CLIENTE_CONTATO_EMAILS_QUERY, {
    variables: {
      id: String(clienteId),
      pagination: {
        pageNumber: page.emails,
        pageSize: 10,
      },
    },
    fetchPolicy: 'cache-and-network',
  });

  const enderecos = useQuery<
    ClienteContatoEnderecoData,
    ClienteContatoQueryVariables
  >(CLIENTE_CONTATO_ENDERECOS_QUERY, {
    variables: {
      id: String(clienteId),
      pagination: {
        pageNumber: page.enderecos,
        pageSize: 10,
      },
    },
    fetchPolicy: 'cache-and-network',
  });

  const telefones = useQuery<
    ClienteContatoTelefoneData,
    ClienteContatoQueryVariables
  >(CLIENTE_CONTATO_TELEFONES_QUERY, {
    variables: {
      id: String(clienteId),
      pagination: {
        pageNumber: page.telefones,
        pageSize: 10,
      },
    },
    fetchPolicy: 'cache-and-network',
  });

  const titulos = useQuery<TituloData, TitulosQueryVariables>(
    CLIENTE_TITULOS_QUERY,
    {
      variables: {
        clienteId: clienteId || '',
      },
      fetchPolicy: 'cache-and-network',
    },
  );

  const clienteFollowUps = useQuery<
    ClienteFollowUpsData,
    ClienteFollowUpsQueryVariables
  >(CLIENTE_FOLLOWUPS_QUERY, {
    variables: {
      clienteId: String(clienteId),
      pagination: {
        pageNumber: 1,
        pageSize: 10,
      },
    },
    fetchPolicy: 'cache-and-network',
    onCompleted(data) {
      if (data.followUps.pageInfo.hasNextPage) {
        clienteFollowUps.refetch({
          pagination: {
            pageNumber: 1,
            pageSize: data.followUps.pageInfo.totalItems || 10,
          },
        });
      }
    },
  });

  const situacoesFollowUp = useQuery<
    SituacoesFollowUpData,
    SituacoesFollowUpsQueryVariables
  >(SITUACOES_FOLLOWUP_QUERY, {
    variables: {
      keyword: situacoesKeyword,
      pagination: {
        pageNumber: 1,
        pageSize: 10,
      },
    },
    onCompleted(data) {
      if (data.situacoesFollowUp.pageInfo.hasNextPage) {
        situacoesFollowUp.refetch({
          pagination: {
            pageNumber: 1,
            pageSize: data.situacoesFollowUp.pageInfo.totalItems || 10,
          },
        });
      }
    },
    fetchPolicy: 'cache-and-network',
  });

  const refetch = (
    target: 'emails' | 'telefones' | 'enderecos' | 'titulos',
  ) => {
    const query = { emails, telefones, enderecos, titulos };
    setPage({ [target]: 1 });

    query[target].refetch({
      id: String(clienteId),
      pagination: {
        pageNumber: 1,
        pageSize: 10,
      },
    });
  };

  const loadMore = (target: 'emails' | 'telefones' | 'enderecos') => {
    const query = {
      emails,
      telefones,
      enderecos,
    };

    query[target].fetchMore({
      variables: {
        pagination: {
          pageNumber: page[target] + 1,
          pageSize: 10,
        },
      },
      updateQuery: (prev, { fetchMoreResult }) => {
        if (!fetchMoreResult) return prev;

        return deepMerge(prev, fetchMoreResult);
      },
    });

    setPage((page) => ({ [target]: Number(page[target]) + 1 }));
  };

  useEffect(() => {
    if (enderecos.data) {
      setContato({
        enderecos: enderecos.data.enderecos,
      });
    }
    if (emails.data) {
      setContato({
        emails: emails.data.emails,
      });
    }
    if (telefones.data) {
      setContato({
        telefones: telefones.data.telefones,
      });
    }
  }, [telefones.data, emails.data, enderecos.data]);

  useEffect(() => {
    situacoesFollowUp.refetch({
      keyword: situacoesKeyword,
      pagination: {
        pageNumber: 1,
        pageSize:
          situacoesFollowUp.data?.situacoesFollowUp.pageInfo.totalItems || 10,
      },
    });
  }, [situacoesKeyword]);

  const isLoading = [
    loading,
    emails.loading,
    enderecos.loading,
    telefones.loading,
    titulos.loading,
  ].includes(true);

  return {
    followps: {
      items: clienteFollowUps.data?.followUps.items.map(
        (followUp) => followUp.node,
      ),
      situacoes: situacoesFollowUp.data?.situacoesFollowUp.items.map(
        (situacao) => situacao.node,
      ),
      setSituacoesKeyword,
    },
    titulo: titulos.data?.titulos.items.at(0)?.node || null,
    cliente: data?.cliente
      ? {
          ...data?.cliente,
          contato: contato as Required<Contato>,
        }
      : undefined,
    loading: isLoading,
    loadMore,
    refetch,
    route,
    openDashboard,
  };
}
