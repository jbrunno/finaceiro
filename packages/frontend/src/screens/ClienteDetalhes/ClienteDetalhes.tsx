import { ClienteContatoCard } from '@/shared/components/ClienteContato/ClienteContatoCard';
import { ClienteInfoCard } from '@/shared/components/ClienteInfoCard/ClienteInfoCard';
import { ClienteHeader } from '@/shared/components/Headers/ClienteHeader/ClienteHeader';
import {
  Breadcrumbs,
  Link,
  RenderComponents,
  Typography,
} from '@frontend/components';
import { Content, Widgets, WidgetsColumn } from '@frontend/styles';
import { NavigateNext as NavigateNextIcon } from '@mui/icons-material';
import { TituloCard } from '@/shared/components/TituloCard/TituloCard';
import { FollowUpCard } from '@/shared/components/FollowUpCard/FollowUpCard';
import { useClienteDetalhes } from './hooks/useClienteDetalhes';

export function ClienteDetalhes() {
  const {
    cliente,
    titulo,
    followps,
    loading,
    loadMore,
    refetch,
    route,
    openDashboard,
  } = useClienteDetalhes();

  const atendimento = {
    id: '1',
    cliente: {
      id: cliente?.id || '1',
    },
    contatoTelefone: undefined,
  };

  return (
    <Content>
      <ClienteHeader
        cliente={cliente}
        navigationBack={openDashboard}
        breadcrumbs={
          <Breadcrumbs separator={<NavigateNextIcon />}>
            <Link variant="body1" color="textSecondary" href={route}>
              Dashboard
            </Link>
            <Typography variant="body1" color="textPrimary">
              Perfil do cliente
            </Typography>
          </Breadcrumbs>
        }
      />
      <Widgets>
        <WidgetsColumn
          sx={{
            gridColumn: '1 / 5',
          }}
        >
          <RenderComponents screen="clienteDetalhes">
            <ClienteInfoCard
              screen="clienteDetalhes"
              cliente={cliente}
              contato={atendimento.contatoTelefone}
              perm="CLIENTE_INFO_CARD"
            />
            <ClienteContatoCard
              loading={loading}
              loadMore={loadMore}
              refetch={refetch}
              cliente={cliente}
              atendimento={atendimento}
              perm="CLIENTE_CONTATO_CARD"
            />
          </RenderComponents>
        </WidgetsColumn>
        <WidgetsColumn
          sx={{
            gridColumn: '5 / 9',
          }}
        >
          <RenderComponents screen="clienteDetalhes">
            <TituloCard
              screen="clienteDetalhes"
              perm="CLIENTE_TITULO_CARD"
              loading={loading}
              assinatura={titulo?.assinatura}
              refetch={refetch}
              titulo={titulo}
            />
            <FollowUpCard
              tituloId={titulo?.id}
              followps={followps}
              perm="CLIENTE_HISTORICO_CARD"
            />
          </RenderComponents>
        </WidgetsColumn>
      </Widgets>
    </Content>
  );
}
