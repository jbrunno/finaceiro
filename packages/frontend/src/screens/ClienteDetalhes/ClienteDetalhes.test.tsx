import { render } from '@test-utils/component-testing';
import { ClienteDetalhes } from './ClienteDetalhes';
import {
  clienteFollowUps,
  cliente as clienteMock,
  emails,
  enderecos,
  situacoesFollowUp,
  telefones,
  titulos,
} from './hooks/useClienteDetalhes.mock';
import * as hook from './hooks/useClienteDetalhes';

jest.mock('@frontend/hooks', () => ({
  ...jest.requireActual('@frontend/hooks'),
  useObjectState: (initial: any) => [initial, jest.fn],
}));

jest.spyOn(hook, 'useClienteDetalhes').mockImplementation(() => ({
  loading: false,
  route: '/test',
  openDashboard: jest.fn,
  titulo: titulos.items[0].node,
  cliente: {
    ...clienteMock,
    contato: {
      emails,
      enderecos,
      telefones,
    },
  },
  followps: {
    items: clienteFollowUps.items.map((followUp) => followUp.node),
    situacoes: situacoesFollowUp.items.map((situacao) => situacao.node),
    setSituacoesKeyword: jest.fn,
  },
  loadMore: jest.fn,
  refetch: jest.fn,
}));

test('should render ClienteDetalhes', () => {
  const { container, rerender } = render(<ClienteDetalhes />);

  rerender(<div />);
  expect(container).toMatchSnapshot();
});
