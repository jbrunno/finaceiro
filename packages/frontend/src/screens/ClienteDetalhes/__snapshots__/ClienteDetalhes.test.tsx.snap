// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ClienteDetalhes 1`] = `
<div>
  <div />
</div>
`;
